index.rst

```diff
@@ -13,17 +13,542 @@
+Most prominent of these is the :meth:`PipelineWorkspaceButler.run_quanta` method, which guarantees correct behavior in the presence of concurrent calls as long as those calls identify disjoint sets of quanta to run, allowing it to be used as an execution primitive by higher-level logic capable of generating those disjoint sets (e.g. BPS).
+
+In fact, we will define :meth:`PipelineWorkspaceButler.run_quanta` to *block* when it is given a quantum whose inputs are produced by other quanta that have not been run.
+This is possible because this workspace implementation will also record quantum status (e.g. in per-quantum report files) for provenance, and we can modify the single-quantum execution logic to wait (e.g. for these files) to appear upstream.
+This actually addresses a possible soundness problem in our current execution model: :class:`~lsst.resources.ResourcePath` backends other than POSIX and S3 may not guarantee that an output file artifacts write will propagate to all servers by the time a read or existence check is performed on a downstream node, let alone by the time the write call returns, and we think we've seen this problem in some WebDAV storage in the past.
+With our current approach to execution this can result in silently incorrect execution, because the dataset will be assumed to have been not produced and may be ignored by the downstream.
+
+Phasing out prediction-mode datastore
+-------------------------------------
+
+Allowing file datastore to operate in "prediction mode" (where a datastore client assumes datasets were written with the same formatters and file templates it is configured to use) was a necessary part of getting QBB working.
+With workspaces, we hope to retire it, by saving datastore records along with the quantum status information the workspace will already be required to store, and which downstream quanta will be required to read.
+This will allow us to simplify the datastore code and eliminate existence checks both during execution and when committing a quantum-graph-based workflow back to the central data repository.
+
+Pipeline workspace actions
+--------------------------
+
+Creating a new pipeline workspace will not require that the quantum graph be produced immediately, though this will be a convenience operation (combining all of the steps described below) we expect to be frequently exercised.
+Instead, only the pipeline must be provided up-front, and even that can be overwritten in development mode.
+Tasks within that pipeline are activated next, which writes their init output file artifacts to the workspace (note that this now happens before a quantum graph is built).
+Quanta can then be built, shard-by-shard; the workspace will remember which shards have quanta already.
+Even executed quanta are not immutable until the workspace is committed - :class:`PipelineWorkspaceButler` will provide methods for quantum state transitions, like resetting a failed quantum so it can be attempted again.
```

Comment:
Do you mean quanta state (provenance)? I don't think we expect quanta themselves to be modified?

Reply:
Yes, though we might want to adjust our nomenclature to allow the word "quantum" to encompass its status, if that makes it easier to talk about provenance.

And resetting a quantum's status _is_ the point that I think it sometimes makes sense to delete its existing outputs, so that could happen, too.  Clobbering only when we actually write a replacement file has always been a bit worrying and messy to me.